src/python/pants/backend/go/util_rules/cgo.py

```diff
@@ -0,0 +1,858 @@
+    cxxflags: tuple[str, ...]
+    fflags: tuple[str, ...]
+    ldflags: tuple[str, ...]
+    pkg_config: tuple[str, ...]
+
+
+@dataclass(frozen=True)
+class CheckCompilerSupportsFlagRequest:
+    cc: str
+    flag: str
+
+
+@dataclass(frozen=True)
+class CheckCompilerSupportsOptionResult:
+    supports_flag: bool
+
+
+# Logic and comments in this rule come from `go` at:
+# https://github.com/golang/go/blob/7eaad60737bc507596c56cec4951b089596ccc9e/src/cmd/go/internal/work/exec.go#L2570
+@rule
+async def check_compiler_supports_flag(
+    request: CheckCompilerSupportsFlagRequest, goroot: GoRoot
+) -> CheckCompilerSupportsOptionResult:
+    input_digest = EMPTY_DIGEST
+    tmp_file = "/dev/null"
+    if goroot.goos == "windows":
+        input_digest = await Get(Digest, CreateDigest([FileContent("grok.c", b"")]))
+        tmp_file = "grok.c"
+
+    # We used to write an empty C file, but that gets complicated with
+    # go build -n. We tried using a file that does not exist, but that
+    # fails on systems with GCC version 4.2.1; that is the last GPLv2
+    # version of GCC, so some systems have frozen on it.
+    # Now we pass an empty file on stdin, which should work at least for
+    # GCC and clang.
+
+    result = await Get(
+        FallibleProcessResult,
+        Process(
+            [request.cc, request.flag, "-c", "-x", "c", "-", "-o", tmp_file],
+            input_digest=input_digest,
+            env={
+                "LC_ALL": "C",
+            },
+            description=f"Check whether CC supports flag: {request.flag}",
+        ),
+    )
+
+    # GCC says "unrecognized command line option".
+    # clang says "unknown argument".
+    # Older versions of GCC say "unrecognised debug output level".
+    # For -fsplit-stack GCC says "'-fsplit-stack' is not supported".
+    combined_output = result.stdout + result.stderr
+    supported = (
+        b"unrecognized" not in combined_output
+        and b"unknown" not in combined_output
+        and b"unrecognised" not in combined_output
+        and b"is not supported" not in combined_output
+    )
+    return CheckCompilerSupportsOptionResult(supported)
+
+
+@dataclass(frozen=True)
+class SetupCompilerCmdRequest:
+    compiler: tuple[str, ...]
+    include_dir: str
+
+
+@dataclass(frozen=True)
+class SetupCompilerCmdResult:
+    args: tuple[str, ...]
+
+
+def _gcc_arch_args(goroot: GoRoot) -> list[str]:
+    goarch = goroot.goarch
+    if goarch == "386":
+        return ["-m32"]
+    elif goarch == "amd64":
+        if goroot.goos == "darwin":
+            return ["-arch", "x86_64", "-m64"]
+        return ["-m64"]
+    elif goarch == "arm64":
+        if goroot.goos == "darwin":
+            return ["-arch", "arm64"]
+    elif goarch == "arm":
+        return ["-marm"]  # not thumb
+    elif goarch == "s390x":
+        return ["-m64", "-march=z196"]
+    elif goarch in ("mips64", "mips64le"):
+        args = ["-mabi=64"]
+        # if cfg.GOMIPS64 == "hardfloat" {
```

Comment:
Do we want to uncomment or delete these? I'm usually leery of commented-out code, unless it's temporary, or explained.

Reply:
Say all that in a comment, and leave it in :)